torchio/data/sampler/weighted.py

```diff
@@ -0,0 +1,262 @@
+        >>> for patch in sampler(sample):
+        ...     print(patch['index_ini'])
+
+    .. note:: The index of the center of a patch with even size :math:`s` is
+        arbitrarily set to :math:`s/2`. This is an implementation detail that
+        will typically not make any difference in practice.
+
+    .. note:: Values of the probability map near the border will be set to 0 as
+        the center of the patch cannot be at the border (unless the patch has
+        size 1 or 2 along that axis).
+
+    """
+    def __init__(
+            self,
+            patch_size: TypePatchSize,
+            probability_map: Optional[str] = None,
+            ):
+        super().__init__(patch_size)
+        self.probability_map_name = probability_map
+        self.cdf = None
+        self.sort_indices = None
+
+    def __call__(
+            self,
+            sample: Subject,
+            num_patches: Optional[int] = None,
+            ) -> Generator[Subject, None, None]:
+        sample.check_consistent_shape()
+        if np.any(self.patch_size > sample.spatial_shape):
+            message = (
+                f'Patch size {tuple(self.patch_size)} cannot be'
+                f' larger than image size {tuple(sample.spatial_shape)}'
+            )
+            raise RuntimeError(message)
+        probability_map = self.get_probability_map(sample)
+        probability_map = self.process_probability_map(probability_map)
+        cdf, sort_indices = self.get_cumulative_distribution_function(
+            probability_map)
+
+        patches_left = num_patches if num_patches is not None else True
+        while patches_left:
+            yield self.extract_patch(sample, probability_map, cdf, sort_indices)
+            if num_patches is not None:
+                patches_left -= 1
+
+    def get_probability_map(self, sample: Subject) -> torch.Tensor:
+        if self.probability_map_name in sample:
+            data = sample[self.probability_map_name].data
+        else:
+            message = (
+                f'Image "{self.probability_map_name}"'
+                f' not found in subject sample: {sample}'
+            )
+            raise KeyError(message)
+        if torch.any(data < 0):
+            message = (
+                'Negative values found'
+                f' in probability map "{self.probability_map_name}"'
+            )
+            raise ValueError(message)
+        return data
+
+    def process_probability_map(
+            self,
+            probability_map: torch.Tensor,
+            ) -> np.ndarray:
+        # Using float32 can create cdf with maximum very far from 1, e.g. 0.92!
+        data = probability_map[0].numpy().astype(np.float64)
+        assert data.ndim == 3
+        if data.sum() == 0:  # although it should not be empty
+            data += 1  # make uniform
+        data /= data.sum()  # normalize probabilities
+        self.clear_probability_borders(data, self.patch_size)
+        assert data.sum() > 0
```

Comment:
I've been wondering that, I'm not sure. Why do you think it would be better?

Reply:
Well you both normalize and clear borders but after clearing borders, if there was a non zero probability on the border your probabilities are not normalized anymore so the previous step becomes useless. Furthermore you are normalizing again when you generate the cdf. That's why I think it would be better clearing borders first and normalizing after (without forgetting to check that `data.sum() > 0` between those 2 steps).

Reply:
I think I remember why I ended up choosing this. What do you think should happen if the array is all zeros after clearing the borders? I guess there are two possibilities:
1. Raise a warning, fill with ones (make uniform), clear borders again
2. Raise an error

Reply:
I don't know, I think I will raise an error because it seems like a really specific probability map was given and a uniform one would be far from what was initially given...